### ISSUE TITLE
Add a lightweight/minimal layer version?

### ISSUE THREAD

**Issue by shouze** [brefphp/bref#722]
Hi,

as always deciding about what to put or not into php during compile time is a matter of arbitrary decisions.
Event if I'm comfortable with most of them, I wonder if it would be a good idea to create more layer variants, especially a minimal one.

I've cloned this repo and made a quick test on php-74-fpm.
By removing lot of things I obtain a **20% smaller layer**.

### Before

```sh
bash-4.2$ du -sh /opt/bref
68M     /opt/bref
bash-4.2$ du -sh /opt/bref/bin/php 
15M     /opt/bref/bin/php
``` 

### After

```sh
bash-4.2$ du -sh /opt/bref
55M     /opt/bref
bash-4.2$ du -sh /opt/bref/bin/php 
14M     /opt/bref/bin/php
``` 

I've also seen that both shared (*.so) and static (*.a) extensions versions are in the produced layer:
```sh
bash-4.2$ ls -lha /opt/bref/lib/php/extensions/no-debug-non-zts-20190902/
total 7.6M
drwxr-xr-x 2 root root  4.0K Aug 19 08:59 .
drwxr-xr-x 3 root root  4.0K Aug 19 08:58 ..
-rw-r--r-- 1 root root  557K Aug 19 08:59 apcu.so
-rwxr-xr-x 1 root root  1.2M Aug 19 08:59 intl.a
-rwxr-xr-x 1 root root  1.6M Aug 19 08:58 intl.so
-rwxr-xr-x 1 root root 1008K Aug 19 08:59 opcache.a
-rwxr-xr-x 1 root root  590K Aug 19 08:58 opcache.so
-rwxr-xr-x 1 root root   50K Aug 19 08:59 pdo_mysql.a
-rwxr-xr-x 1 root root   34K Aug 19 08:58 pdo_mysql.so
-rwxr-xr-x 1 root root   77K Aug 19 08:59 pdo_pgsql.a
-rwxr-xr-x 1 root root   54K Aug 19 08:58 pdo_pgsql.so
-rw-r--r-- 1 root root  2.6M Aug 19 08:59 redis.so
```

I have tested yet but I'm not sure we also need them? :thinking: 
If it's the case the cleanup script could remove them with at least 2M gain.

I haven't deployed this layer nor tested on true lambda invocations yet but as the produced layer is smaller I guess performance should be better (especially cold starts).

**Comment by shouze** [comment]
I forgot to mention the resulting size of the exported layer:
```sh 
❯ ls -lha export/php-74-fpm.zip 
Permissions Size User      Date Modified  Git Name
.rw-rw-r--   30M sebastien 19 août  11:25  -I  export/php-74-fpm.zip
```

**Comment by shouze** [comment]
And here's the corresponding Dockerfile:
```Dockerfile
# The container we build here contains everything needed to compile PHP + PHP.
#
# It can be used as a base to compile extra extensions.

# PHP Build
# https://github.com/php/php-src/releases
# Needs:
#   - zlib
#   - openssl
#   - readline
#   - sodium

FROM bref/tmp/step-1/build-environment as build-environment


###############################################################################
# Oniguruma
# This library is not packaged in PHP since PHP 7.4.
# See https://github.com/php/php-src/blob/43dc7da8e3719d3e89bd8ec15ebb13f997bbbaa9/UPGRADING#L578-L581
# We do not install the system version because I didn't manage to make it work...
# Ideally we shouldn't compile it ourselves.
# https://github.com/kkos/oniguruma/releases
# Needed by:
#   - php mbstring
ENV VERSION_ONIG=6.9.3
ENV ONIG_BUILD_DIR=${BUILD_DIR}/oniguruma
RUN set -xe; \
    mkdir -p ${ONIG_BUILD_DIR}; \
    curl -Ls https://github.com/kkos/oniguruma/releases/download/v${VERSION_ONIG}/onig-${VERSION_ONIG}.tar.gz \
    | tar xzC ${ONIG_BUILD_DIR} --strip-components=1
WORKDIR  ${ONIG_BUILD_DIR}/
RUN set -xe; \
    ./configure --prefix=${INSTALL_DIR}; \
    make -j $(nproc); \
    make install


ENV VERSION_PHP=7.4.9


ENV PHP_BUILD_DIR=${BUILD_DIR}/php
RUN set -xe; \
    mkdir -p ${PHP_BUILD_DIR}; \
    # Download and upack the source code
    # --location will follow redirects
    # --silent will hide the progress, but also the errors: we restore error messages with --show-error
    # --fail makes sure that curl returns an error instead of fetching the 404 page
    curl --location --silent --show-error --fail https://www.php.net/get/php-${VERSION_PHP}.tar.gz/from/this/mirror \
    | tar xzC ${PHP_BUILD_DIR} --strip-components=1
# Move into the unpackaged code directory
WORKDIR  ${PHP_BUILD_DIR}/

# Configure the build
# -fstack-protector-strong : Be paranoid about stack overflows
# -fpic : Make PHP's main executable position-independent (improves ASLR security mechanism, and has no performance impact on x86_64)
# -fpie : Support Address Space Layout Randomization (see -fpic)
# -O3 : Optimize for fastest binaries possible.
# -I : Add the path to the list of directories to be searched for header files during preprocessing.
# --enable-option-checking=fatal: make sure invalid --configure-flags are fatal errors instead of just warnings
# --enable-ftp: because ftp_ssl_connect() needs ftp to be compiled statically (see https://github.com/docker-library/php/issues/236)
# --enable-mbstring: because otherwise there's no way to get pecl to use it properly (see https://github.com/docker-library/php/issues/195)
# --with-zlib and --with-zlib-dir: See https://stackoverflow.com/a/42978649/245552
# --with-pear: necessary for `pecl` to work (to install PHP extensions)
#
RUN set -xe \
    && ./buildconf --force \
    && CFLAGS="-fstack-protector-strong -fpic -fpie -O3 -I${INSTALL_DIR}/include -I/usr/include -ffunction-sections -fdata-sections" \
    CPPFLAGS="-fstack-protector-strong -fpic -fpie -O3 -I${INSTALL_DIR}/include -I/usr/include -ffunction-sections -fdata-sections" \
    LDFLAGS="-L${INSTALL_DIR}/lib64 -L${INSTALL_DIR}/lib -Wl,-O1 -Wl,--strip-all -Wl,--hash-style=both -pie" \
    ./configure \
    --build=x86_64-pc-linux-gnu \
    --prefix=${INSTALL_DIR} \
    --enable-option-checking=fatal \
    --enable-sockets \
    --with-config-file-path=${INSTALL_DIR}/etc/php \
    --with-config-file-scan-dir=${INSTALL_DIR}/etc/php/conf.d:/var/task/php/conf.d \
    --enable-fpm \
    --disable-cgi \
    --enable-cli \
    --disable-phpdbg \
    --disable-phpdbg-webhelper \
    --with-sodium \
    --with-readline \
    --with-openssl \
    --with-zlib=${INSTALL_DIR} \
    --with-zlib-dir=${INSTALL_DIR} \
    --with-curl \
    --disable-exif \
    --disable-ftp \
    --enable-mbstring \
    --with-pdo-mysql=shared,mysqlnd \
    --with-mysqli \
    --enable-pcntl \
    --with-zip \
    --enable-bcmath \
    --with-pdo-pgsql=shared,${INSTALL_DIR} \
    --enable-intl=shared \
    --disable-soap \
    --with-xsl=${INSTALL_DIR} \
    --disable-gd \
    --with-pear
RUN make -j $(nproc)
# Run `make install` and override PEAR's PHAR URL because pear.php.net is down
RUN set -xe; \
    make install PEAR_INSTALLER_URL='https://github.com/pear/pearweb_phars/raw/master/install-pear-nozlib.phar'; \
    { find ${INSTALL_DIR}/bin ${INSTALL_DIR}/sbin -type f -perm +0111 -exec strip --strip-all '{}' + || true; }; \
    make clean; \
    cp php.ini-production ${INSTALL_DIR}/etc/php/php.ini

# Symlink all our binaries into /opt/bin so that Lambda sees them in the path.
RUN mkdir -p /opt/bin
RUN ln -s /opt/bref/bin/* /opt/bin
RUN ln -s /opt/bref/sbin/* /opt/bin

# Install extensions
# We can install extensions manually or using `pecl`
RUN pecl install redis
RUN pecl install APCu


# Run the next step in the previous environment because the `clean.sh` script needs `find`,
# which isn't installed by default
FROM build-environment as build-environment-cleaned
# Remove extra files to make the layers as slim as possible
COPY clean.sh /tmp/clean.sh
RUN /tmp/clean.sh && rm /tmp/clean.sh


# Now we start back from a clean image.
# We get rid of everything that is unnecessary (build tools, source code, and anything else
# that might have created intermediate layers for docker) by copying online the /opt directory.
FROM amazonlinux:2018.03
ENV PATH="/opt/bin:${PATH}" \
    LD_LIBRARY_PATH="/opt/bref/lib64:/opt/bref/lib"

# Copy everything we built above into the same dir on the base AmazonLinux container.
COPY --from=build-environment-cleaned /opt /opt

# Set the workdir to the same directory as in AWS Lambda
WORKDIR /var/task
```

**Comment by mnapoli** [comment]
Hey! This is interesting. I'm not really keen on adding another layer, but maybe we can trim the existing one.

Since 1.0 should happen soon, we could revisit [the list of provided extensions](https://bref.sh/docs/environment/php.html#extensions) and do some cleanup, since some of them can be installed via the "bref extra extensions".

For example, maybe gd, imagick, and also maybe mongodb could go? It may be worth looking into which extensions weight the most.

**Comment by shouze** [comment]
I'm agree that it would be better to maintain only a thin layer but I didn't suggest that as it's not BC. Would be nice indeed for `1.0` if no fear about BC break prior to 1st stable release.

And about `.a` files, do you think we can remove them?

How can I help anyway? To help 1.0 release I've identified following tasks to achieve the goal:
- [ ] Identify biggest extensions
- [ ] Remove them from base image
- [ ] Provide them in separate [layers/extra extensions](https://github.com/brefphp/extra-php-extensions)
- [ ] Update documentation accordingly

Feel free to open those issues and any others, I think I can help/assign me for at least 50% of them and work on it next week :wink: 

As a bonus, maybe using [buildx](https://github.com/docker/buildx) could be a build time booster for ci/cd. A [Github Action](https://github.com/crazy-max/ghaction-docker-buildx) is available in case. I've tested it yesterday because I (still) use docker as a package builder for my lambda and I wanted to find a way to not deploy my lambdas each time from ci/cd.

**Comment by shouze** [comment]
Ha, also what about ftp support? We should keep it, I've disabled it at the moment but [after looking at official alpine docker configure](https://github.com/docker-library/php/blob/master/7.4/alpine3.12/fpm/Dockerfile#L122-L159) I'm pretty condident that we should do almost the same, WDYT? :thinking:

**Comment by shouze** [comment]
# What to remove?

For the record (latest php-74-fpm here):
```sh
bash-4.2# du -sh /opt/bref
66M     /opt/bref
```

**We can save about 23M**, about **35% saving over the current layer** (uncompressed) size :package: !
Here's how we can proceed.

## Static libraries & extensions

We can remove them all being careful (especially with [sodium](https://github.com/docker-library/php/blob/098e442542e8a10bdee2c22484a98d41583a8fb9/7.4/buster/apache/Dockerfile#L217)). This is [done in official php docker images](https://github.com/docker-library/php/blob/098e442542e8a10bdee2c22484a98d41583a8fb9/7.4/buster/apache/Dockerfile#L238). So I guess we can also do it to **save 11M**.

```sh
bash-4.2# find /opt/bref/ -type f -name '*.a' | xargs du -ch 
996K    /opt/bref/lib64/libssl.a
5.3M    /opt/bref/lib64/libcrypto.a
1.1M    /opt/bref/lib/php/extensions/no-debug-zts-20190902/opcache.a
80K     /opt/bref/lib/php/extensions/no-debug-zts-20190902/pdo_pgsql.a
52K     /opt/bref/lib/php/extensions/no-debug-zts-20190902/pdo_mysql.a
1.2M    /opt/bref/lib/php/extensions/no-debug-zts-20190902/intl.a
560K    /opt/bref/lib/libsodium.a
732K    /opt/bref/lib/libonig.a
328K    /opt/bref/lib/libpq.a
11M     total
```

## Biggest/useless extensions

I suggest to keep only opcache & apcu by default. With the 5 layers limit you can add up to 4 extensions if you use a single layer each time for each extension, but it should also worth to document/think about bundling all extra extensions in a single layer, WDYT? If you're agree with that we can **save here an extra 12M**.

```sh
bash-4.2# find /opt/bref/lib/php/extensions/no-debug-zts-20190902 -type f -name '*.so' | xargs du -ch
6.0M    /opt/bref/lib/php/extensions/no-debug-zts-20190902/mongodb.so
2.5M    /opt/bref/lib/php/extensions/no-debug-zts-20190902/redis.so
1.5M    /opt/bref/lib/php/extensions/no-debug-zts-20190902/intl.so
1.2M    /opt/bref/lib/php/extensions/no-debug-zts-20190902/imagick.so
616K    /opt/bref/lib/php/extensions/no-debug-zts-20190902/opcache.so
488K    /opt/bref/lib/php/extensions/no-debug-zts-20190902/apcu.so
56K     /opt/bref/lib/php/extensions/no-debug-zts-20190902/pdo_pgsql.so
36K     /opt/bref/lib/php/extensions/no-debug-zts-20190902/pdo_mysql.so
13M     total
```

### Other things

- gd
- maintainer-zts  & pthreads (as already suggested by @Nyholm in https://github.com/brefphp/bref/pull/702)

**Comment by shouze** [comment]
Would be nice also to be able at some point to replace libreadline by libedit (more modern). As l libreadline is provided by aws base  layer and while libedit is not, this imply to compile and add libedit in bref layer.

Also would be great to add argon2 passwords as recommanded in modern php.

**Comment by shouze** [comment]
closed by #775
